File: typescript/packages/management/src/tools/provision_source.ts
import type { CreateDataSourceRequest, DataSourceIngestionMode, DataSourceKind } from '@copass/core';

import type { ToolContext, ToolHandler } from '../registrar.js';

export const provisionSource: ToolHandler = async (
  ctx: ToolContext,
  input: Record<string, unknown>,
) => {
  const adapterConfig: Record<string, unknown> =
    typeof input.adapter_config === 'object' && input.adapter_config !== null
      ? { ...(input.adapter_config as Record<string, unknown>) }
      : {};
  if (input.ingest_to_graph === true) {
    adapterConfig.ingest_to_graph = true;
  }

  const request: CreateDataSourceRequest = {
    provider: typeof input.provider === 'string' ? input.provider : 'pipedream',
    name: String(input.name),
  };
  if (typeof input.ingestion_mode === 'string') {
    request.ingestion_mode = input.ingestion_mode as DataSourceIngestionMode;
  }
  if (input.kind === 'durable' || input.kind === 'ephemeral') {
    request.kind = input.kind as DataSourceKind;
  }
  if (typeof input.external_account_id === 'string') {
    request.external_account_id = input.external_account_id;
  }
  if (typeof input.poll_interval_seconds === 'number') {
    request.poll_interval_seconds = input.poll_interval_seconds;
  }
  if (Object.keys(adapterConfig).length > 0) {
    request.adapter_config = adapterConfig;
  }

  const source = await ctx.client.sources.register(ctx.sandboxId, request);
  return { source };
};

File: python/copass-management/src/copass_management/tools/provision_source.py
from __future__ import annotations

from typing import Any, Dict, TYPE_CHECKING

if TYPE_CHECKING:
    from copass_management.registrar import ToolContext


async def provision_source(ctx: "ToolContext", input: Dict[str, Any]) -> Any:
    adapter_config: Dict[str, Any] = {}
    if isinstance(input.get("adapter_config"), dict):
        adapter_config.update(input["adapter_config"])
    if input.get("ingest_to_graph") is True:
        adapter_config["ingest_to_graph"] = True

    kwargs: Dict[str, Any] = {
        "provider": str(input.get("provider") or "pipedream"),
        "name": str(input["name"]),
    }
    if input.get("ingestion_mode") is not None:
        kwargs["ingestion_mode"] = str(input["ingestion_mode"])
    if input.get("kind") is not None:
        kwargs["kind"] = str(input["kind"])
    if input.get("external_account_id") is not None:
        kwargs["external_account_id"] = str(input["external_account_id"])
    if input.get("poll_interval_seconds") is not None:
        kwargs["poll_interval_seconds"] = int(input["poll_interval_seconds"])
    if adapter_config:
        kwargs["adapter_config"] = adapter_config

    source = await ctx.client.sources.register(ctx.sandbox_id, **kwargs)
    return {"source": source}
